### PR TITLE
Fix service endpoint address validation

### LINKED ISSUE #9743: ServiceEntry update not regenerating config
**Describe the bug**
This may be a more general regen problem, but my particular example is that I've created a ServiceEntry that redirects port 80 and 443 to an egressgateway (config with ISTIO_META_REQUESTED_NETWORK_VIEW=external) using an external network endpoint. This works as expected, but when I later remove one of the ports from the ServiceEntry and apply, the generated proxy config does not change.

**Expected behavior**
I left it running for 2 days, and still the update never happened, so it seems this is not just a long propagation delay issue.

**Steps to reproduce the bug**
Apply the following configuration.
```
kubectl apply -f - <<EOF
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-egressgateway
spec:
  selector:
    istio: egressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - edition.cnn.com
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - edition.cnn.com
    tls:
      mode: PASSTHROUGH
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: edition-cnn-com
spec:
  hosts:
  - edition.cnn.com
  resolution: DNS
  endpoints:
  - address: istio-egressgateway.istio-system.svc.cluster.local
    ports:
      https: 443
      http: 80
  - address: edition.cnn.com
    network: external
    ports:
      https: 443
      http: 80
  ports:
  - number: 443
    name: https
    protocol: HTTPS
  - number: 80
    name: http
    protocol: HTTP
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: edition-cnn-com
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - istio-egressgateway
  tls:
  - match:
    - gateways:
      - istio-egressgateway
      port: 443
      sni_hosts:
      - edition.cnn.com
    route:
    - destination:
        host: edition.cnn.com
        port:
          number: 443
  http:
  - match:
    - gateways:
      - istio-egressgateway
      port: 80
    route:
    - destination:
        host: edition.cnn.com
        port:
          number: 80
EOF
```
Using the sleep sample
```
$ kubectl exec -it $SOURCE_POD -c sleep -- curl -sL -o /dev/null -D - http://edition.cnn.com/politics
```
If you look at the egressgateway logs you will see both http and https calls go through it, as expected.

Now remove the one of the endpoint ports from the serviceentry, e.g.,
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: edition-cnn-com
spec:
  hosts:
  - edition.cnn.com
  resolution: DNS
  endpoints:
  - address: istio-egressgateway.istio-system.svc.cluster.local
    ports:
      #https: 443
      http: 80
  - address: edition.cnn.com
    network: external
    ports:
      #https: 443
      http: 80
  ports:
  - number: 443
    name: https
    protocol: HTTPS
  - number: 80
    name: http
    protocol: HTTP
```
Retry the curl and nothing has changed. 

Look at the generated config in the sidecar, and you will see that it has not changed.

**Version**
Latest from release-1.1 branch

**Installation**

**Environment**
{{ Which environment, cloud vendor, OS, etc are you using? }}

**Cluster state**
{{ If you're running on Kubernetes, consider [following the
instructions](http://istio.io/help/bugs/#generating-a-cluster-state-archive)
to generate "istio-dump.tar.gz", then attach it here by dragging and dropping
the file onto this issue. }}

### PR DESCRIPTION
**Pull request by frankbu**
Fixes #9743

### CODE REVIEW

File: pilot/pkg/model/validation.go [hunk 1]
```diff
@@ -214,7 +214,10 @@ func (l Labels) Validate() error {
 
 // ValidateFQDN checks a fully-qualified domain name
 func ValidateFQDN(fqdn string) error {
-	return appendErrors(checkDNS1123Preconditions(fqdn), validateDNS1123Labels(fqdn))
+	if err := checkDNS1123Preconditions(fqdn); err != nil {
+		return err
```
Comment (rshriram) [inline pilot/pkg/model/validation.go]: you need to appendErrors(..) because we retain the error stack

File: pilot/pkg/model/validation.go [hunk 2]
```diff
@@ -244,7 +247,12 @@ func checkDNS1123Preconditions(name string) error {
 }
 
 func validateDNS1123Labels(domain string) error {
-	for _, label := range strings.Split(domain, ".") {
+	parts := strings.Split(domain, ".")
+	topLevelDomain := parts[len(parts)-1]
+	if len(topLevelDomain) > 0 && '0' <= topLevelDomain[0] && topLevelDomain[0] <= '9' {
```
Comment (rshriram) [inline pilot/pkg/model/validation.go]: can we use the regex on top of this file?

Comment (rshriram) [reply pilot/pkg/model/validation.go]: Also are you sure this is per dns spec that tld cannot start with a number?

File: pilot/pkg/model/validation.go (hunk 1 above)
Comment (frankbu) [reply pilot/pkg/model/validation.go]: You only need appendErrors here if it proceeds to call validateDNS1123Labels, but I don't see any value in doing further checking if the name is already known to be invalid.

File: pilot/pkg/model/validation.go (hunk 2 above)
Comment (frankbu) [reply pilot/pkg/model/validation.go]: https://tools.ietf.org/html/rfc3696#section-2

Not sure what it really means for validation though. The more important question is where in the networking model to we want to allow only DNS names and not IPs. If the answer is nowhere, then I guess the old ValidateFQDN function is fine, and we should clean up any places in the code that call it and also redundantly check for a valid IP

File: pilot/pkg/model/validation.go [hunk 3]
```diff
@@ -1338,7 +1346,10 @@ func ValidateVirtualService(name, namespace string, msg proto.Message) (errs err
 	allHostsValid := true
 	for _, host := range virtualService.Hosts {
 		if err := ValidateWildcardDomain(host); err != nil {
-			errs = appendErrors(errs, err)
+			ipAddr := net.ParseIP(host)
+			if ipAddr == nil { // Could be an IP address
+				errs = appendErrors(errs, err)
```
Comment (rshriram) [inline pilot/pkg/model/validation.go]: what if its not nil? you are still marking allHostsValid=false..

Comment (frankbu) [reply pilot/pkg/model/validation.go]: good catch ... fixed

**Pull request merged**